[Ho Jun Hao] iP

HoJunHao2000 opened the pull request:
# **GigaDuke**
> "It's not a bug, it's a feature" - GigaDuke ([source](https://www.youtube.com/watch?v=dQw4w9WgXcQ&ab_channel=RickAstley))

GigaDuke helps you remember your tasks! it's,

- text-based
- easy to learn
- ~FAST~ SUPER FAST to use

All you need to do is,
1. download it from [here](https://github.com/HoJunHao2000/ip/releases/tag/A-Jar).
2. enter terminal and change to the directory to the one with the file.
3. enter `java -jar ip.jar`
4. let it manage your tasks for you :star_struck:

And it is **FREE**!

~Bugs to fix~ Features:

- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)
-----------------------------------------------------------------------------------------------------------------------
if you are a Java programmer, you can use it to practice Java too. Here's the main() method.

``` java
/**
 * A launcher class to workaround classpath issues.
 */
public class Launcher {
    public static void main(String[] args) {
        Application.launch(Main.class, args);
    }
}
```

Commit:
Add Gradle support

HoJunHao2000 committed:
Level-1

HoJunHao2000 committed:
Level-2

HoJunHao2000 committed:
Level-3

HoJunHao2000 committed:
Level-4

HoJunHao2000 committed:
A-TextUiTesting

HoJunHao2000 committed:
Level-5

HoJunHao2000 committed:
Level-6

HoJunHao2000 committed:
A-Enums

HoJunHao2000 committed:
Add Save and Load functionality

When application restarts, past inputs are erased.

Past inputs would be saved to a file and loaded in when application
restarts.

Users cannot be expected to re-enter past inputs whenever application
restarts.

HoJunHao2000 committed:
Update Deadline and Event input to take in Date and Time

Inputs are in String allowing for invalid inputs.

Input will be restricted to LocalDate type.

To ensure only valid inputs for Date and Time are allowed.

HoJunHao2000 committed:
Merge branch 'branch-Level-8'

* branch-Level-8:
  Update Deadline and Event input to take in Date and Time

# Conflicts:
#	src/main/java/Duke.java

HoJunHao2000 committed:
Update Duke to use OOP

Much of Duke code uses Functional Programming. A switch to OOP would
greatly improve code readability and structure.

Added 4 new classes Ui, Parser, Storage and TaskList to handle various
parts of program.

Added Abstract Command class whose subclasses would run the various
types of commands.

HoJunHao2000 committed:
Divide classes into packages

Currently all the java classes and files are all in a single folder.

Organise the various classes into suitable java packages.

HoJunHao2000 committed:
Merge branch 'add-gradle-support'

* add-gradle-support:
  Add Gradle support

HoJunHao2000 committed:
Add JUnit tests

Add JUnit tests to test the behavior of the code.

Created test cases for adding tasks and exceptions thrown.

HoJunHao2000 committed:
Add JavaDocs to all Classes and methods

Currently, none of the Classes and methods have any description making
it hard for users to understand what each code does.

The addition of JavaDocs would allow for better clarity and
understanding of code.

HoJunHao2000 committed:
Update the code to comply with a coding standard

HoJunHao2000 committed:
Add Find command to Duke

Currently, users have to manually search through all the task to find
the task that they want.

Created Find Command this way users can just input a keyword to see all
tasks with that matching keyword.

HoJunHao2000 committed:
Fix multiple Duke folder error

HoJunHao2000 committed:
Fix incorrect import package statements

HoJunHao2000 committed:
Merge branch 'branch-A-CodingStandard'

* branch-A-CodingStandard:
  Update the code to comply with a coding standard

# Conflicts:
#	src/main/java/duke/command/ByeCommand.java
#	src/main/java/duke/command/ListCommand.java
#	src/main/java/duke/task/TaskList.java
#	src/main/java/duke/ui/DukeException.java
#	src/main/java/duke/ui/Parser.java
#	src/main/java/duke/ui/Ui.java

HoJunHao2000 committed:
Merge branch 'branch-Level-9'

* branch-Level-9:
  Add Find command to Duke

# Conflicts:
#	src/main/java/duke/ui/Parser.java
#	src/main/java/duke/ui/Ui.java

HoJunHao2000 committed:
Add CheckStyle

HoJunHao2000 committed:
Update code format according to checkstyle

HoJunHao2000 committed:
Add CheckStyle to build.gradle

HoJunHao2000 committed:
Fix problem with delete

Currently when deleting, the wrong message gets displayed. Application
is also unable to delete tasks sometimes.

Display message was changed to the right one and bug with list index was
fixed.

HoJunHao2000 committed:
Update JUnit tests to follow CheckStyle format

HoJunHao2000 committed:
Update Duke to use GUI

Currently, all UI is done through terminal which is not visually
appealling.

Usage of JavaFx to create GUI for duke.

HoJunHao2000 committed:
Add Varargs to Find command

Currently, users can only search based on one keyword.

Use varargs, this way users can input multiple keywords to filter in.

HoJunHao2000 committed:
Update JavaDocs for Ui

* Realised mistake of not making branch for Level-10
* Made minor changes for the inclusion of branch-Level-10

HoJunHao2000 committed:
Merge branch 'branch-Level-10'

* branch-Level-10:
  Update JavaDocs for Ui

HoJunHao2000 committed:
Update JavaDocs to meet standards

Currently, many of the JavaDocs lack a line spacing between description
and parameters.

HoJunHao2000 committed:
Fix bug in GUI

Currently, when the list exceeds certain length, it would be cut off
with ...

Edited DialogBox min length to be the dynamic according to length of
list.

bryanljx commented on README.md:
```diff
@@ -1,4 +1,4 @@
-# Duke project template
+# duke.Duke project template
```
This was a mistake that occurred to me too. Using IDE's refactoring functionalities to rename variables can have unintended side effects like these 😅

bryanljx commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,76 @@
+ * Main class for Duke program.
+ * Keeps track of tasks for user.
+ */
+public class Duke {
```
Good to see header comments for public classes 👍 Might want to add the `@author` annotation too to credit yourself!

bryanljx commented on src/main/java/duke/command/AddCommand.java:
```diff
@@ -0,0 +1,43 @@
+     * @throws IOException If there is an error when updating the file.
+     */
+    @Override
+    public String run(TaskList tasks, Ui ui, Storage storage) throws IOException {
+        tasks.addTask(this.task);
+        storage.writeFile(tasks);
+        return ui.printAddTasks(this.task, tasks);
+    }
```
Looks very clean, good job!

bryanljx commented on src/main/java/duke/gui/Main.java:
```diff
@@ -0,0 +1,37 @@
+            Scene scene = new Scene(ap);
+            stage.setScene(scene);
+            fxmlLoader.<MainWindow>getController().setDuke(duke);
+            stage.show();
```
Using fxml really does keep the code short and concise, good job!

bryanljx commented on src/main/java/duke/ui/Parser.java:
```diff
@@ -0,0 +1,116 @@
+     */
+    public static Command parse(String command) throws DukeException, DateTimeParseException {
+        switch (getInput(command)) {
+        case BYE:
+            return new ByeCommand();
+            // Fallthrough
+        case LIST:
+            return new ListCommand();
+            // Fallthrough
+        case MARK:
+            int markNum = Integer.parseInt(command.replace("mark ", ""));
+            return new MarkCommand(markNum - 1);
+            // Fallthrough
+        case UNMARK:
+            int unMarkNum = Integer.parseInt(command.replace("unmark ", ""));
+            return new UnMarkCommand(unMarkNum - 1);
+            // Fallthrough
+        case TODO:
+            String tDes = command.replace("todo", "");
+            return new AddCommand(new Todo(tDes));
+            // Fallthrough
```
This works but having a `// Fallthrough` comment feels redundant since we are already returning here, would removing it be more appropriate?

bryanljx commented on src/main/java/duke/ui/Parser.java:
```diff
@@ -0,0 +1,116 @@
+     * @return An enum that the program understands.
+     */
+    private static Inputs getInput(String command) {
+        if (command.equals("bye")) {
+            return Inputs.BYE;
+        } else if (command.equals("list")) {
+            return Inputs.LIST;
+        } else if (command.startsWith("mark")) {
+            return Inputs.MARK;
+        } else if (command.startsWith("unmark")) {
+            return Inputs.UNMARK;
+        } else if (command.startsWith("todo")) {
+            return Inputs.TODO;
+        } else if (command.startsWith("deadline")) {
+            return Inputs.DEADLINE;
+        } else if (command.startsWith("event")) {
+            return Inputs.EVENT;
+        } else if (command.startsWith("delete")) {
+            return Inputs.DELETE;
+        } else if (command.startsWith("find")) {
+            return Inputs.FIND;
+        } else {
+            return Inputs.ELSE;
+        }
```
Nothing really wrong here but since there are multiple else blocks, perhaps a switch statement is better?

bryanljx reviewed (commented):
Overall very well done and the coding standard is quite strictly adhered too. Keep up the great work 👍

HoJunHao2000 committed:
Update build.grade to include javafx

HoJunHao2000 committed:
Update Duke to throw exceptions when commands given to empty Task list

HoJunHao2000 committed:
Update Duke GUI to display introduction message

HoJunHao2000 committed:
Add assertions to Delete, Mark, Unmark Commands

Users may input invalid indexes after the commands and be unaware of why
command is not executing.

Assertions can help highlight the issues.

Let's,
* Update the relevant Commands to have assertions that highlight failed
preconditions

HoJunHao2000 committed:
Update Duke program to comply with CodeQuality

Some parts of code have complicated expressions.

Having complicated expressions makes it harder for others to trace and
understand the code logic.

Let's,
* Add intermediate values and methods with descriptive names so others
can better understand the code logic.

HoJunHao2000 committed:
Update TaskList to use Streams

Currently, filtering the Task list to Tasks that match the Find
keywords uses a nested for loop.

Usage of streams to demonstrate proficiency.

HoJunHao2000 committed:
Remove dead code found in Ui

HoJunHao2000 committed:
Merge pull request #1 from HoJunHao2000/branch-A-Assertions

Add assertions to Delete, Mark, Unmark Commands

HoJunHao2000 committed:
Merge branch 'master' into branch-A-CodeQuality

HoJunHao2000 committed:
Merge pull request #2 from HoJunHao2000/branch-A-CodeQuality

Update Duke to comply with code quality

HoJunHao2000 committed:
Merge branch 'master' into branch-A-Streams

HoJunHao2000 committed:
Merge pull request #3 from HoJunHao2000/branch-A-Streams

Update TaskList to use Streams

HoJunHao2000 committed:
Add GitHub Actions to set up Continuous Integration (CI)

HoJunHao2000 committed:
Add ability to update the date field of Deadline and Event

Currently, to change the date field, users have to delete and re-enter
the date.

This is inconvenient for users and users cannot be expected to delete
and re-enter every Task.

Let's,
* Add an update function to change date fields for Deadline and Event.

HoJunHao2000 committed:
Update GUI to look better

Currently, the GUI has a plain design with no background or text bubble.
Giving the Bye Command does not close the application too.

Lets update the GUI with
* A background
* Rounded profile pictures
* Bye Command that closes the application

HoJunHao2000 committed:
Add README.md
